harness/tests/tensorboard/test_base.py

```diff
@@ -90,3 +93,63 @@ def test_illegal_type() -> None:
         tensorboard.build(
             env.det_cluster_id, env.det_experiment_id, env.det_trial_id, checkpoint_config
         )
+
+
+def test_upload_thread_normal_case() -> None:
+    upload_function = mock.Mock()
+    upload_thread = tensorboard.base._TensorboardUploadThread(upload_function)
+
+    upload_thread.start()
+    upload_thread.upload(
+        [pathlib.Path("test_value/file1.json"), pathlib.Path("test_value/file2.json")]
+    )
+    upload_thread.upload([pathlib.Path("test_value/file3.json")])
+    # Pass in sentinel value to exit thread
+    upload_thread.close()
+    upload_thread.join()
+
+    assert upload_function.call_count == 2
+
+
+def test_upload_thread_exception_case() -> None:
+    # 1. Define custom hook to capture threads fail
+    #    with exception. This hook is triggered when
+    #    a thread exits with exception.
+    #    Also store orignal excepthook for reset later.
+    import threading
+
+    threads_with_exception = set()
+
+    def custom_excepthook(args):
+        thread_name = args.thread.ident
+        threads_with_exception.add(thread_name)
+
+    original_excepthook = threading.excepthook
```

Comment:
nit: place the saving of the except hook closer to where you override it, to make it more obvious.

Another nit: let the try be the first statement after overwriting it.  This has the cleanest semantics I think, most similar to go's defer statements:

```go
Thing.Open()
# after Thing.Open() returns, the `defer` guarantees we clean it up
defer Thing.Close()
```

Reply:
It turns out excepthook is only available for Python 3.8 and up but our test is run on Python 3.7. 

Without it, I do not think there is a neat way to test this. Using sys.excepthook would not be good as it does not tell you the thread that is causing the exception, other methods would involve actually changing the thread subclass (e.g. pass the exception up in join method) which is not desirable. I will delete this test.